ddtrace/contrib/internal/asgi/middleware.py

```diff
@@ -330,3 +516,228 @@ async def wrapped_blocked_send(message):
+
+            send_span.set_link(
+                trace_id=span.trace_id,
+                span_id=span.span_id,
+                attributes={SPAN_LINK_KIND: SpanLinkKind.RESUMING},
+            )
+            send_span.set_metric(websocket.MESSAGE_FRAMES, 1)
+            _set_message_tags_on_span(send_span, message)
+
+    def _handle_websocket_close_message(self, scope: Mapping[str, Any], message: Mapping[str, Any], span: Span):
+        current_receive_span = scope.get("datadog", {}).get("current_receive_span")
+
+        # Use receive span as parent if it exists, otherwise use main span
+        if current_receive_span:
+            parent_span = current_receive_span
+        else:
+            parent_span = span
+
+        pin = ddtrace.trace.Pin(service=span.service)
+        pin._tracer = self.tracer
+        with core.context_with_data(
+            "asgi.websocket.close_message",
+            pin=pin,
+            tracer=self.tracer,
+            integration_config=self.integration_config,
+            span_name="websocket.close",
+            service=span.service,
+            resource=f"websocket {scope.get('path', '')}",
+            span_type=SpanTypes.WEBSOCKET,
+            child_of=parent_span,
+            call_trace=False,
+            activate=True,
+        ) as ctx, ctx.span as close_span:
+            # with self.tracer.start_span(
+            #     name="websocket.close",
+            #     service=span.service,
+            #     resource=f"websocket {scope.get('path', '')}",
+            #     span_type=SpanTypes.WEBSOCKET,
+            #     child_of=parent_span,
+            #     activate=True,
+            # ) as close_span:
+            close_span.set_tag_str(COMPONENT, self.integration_config.integration_name)
+            close_span.set_tag_str(SPAN_KIND, SpanKind.PRODUCER)
+
+            client = scope.get("client")
+            if len(client) >= 1:
+                client_ip = client[0]
+                close_span.set_tag_str("out.host", client_ip)
+                try:
+                    ipaddress.ip_address(client_ip)  # validate ip address
+                    close_span.set_tag_str("network.client.ip", client_ip)
+                except ValueError:
+                    log.debug("Could not validate client IP address")
+
+            _set_message_tags_on_span(close_span, message)
+            # should act like send span (outgoing message) case
+            close_span.set_link(
+                trace_id=span.trace_id,
+                span_id=span.span_id,
+                attributes={SPAN_LINK_KIND: SpanLinkKind.RESUMING},
+            )
+            _copy_baggage_and_sampling(close_span, span)
+
+            code = message.get("code")
+            reason = message.get("reason")
+            if code is not None:
+                close_span.set_metric(websocket.CLOSE_CODE, code)
+            if reason:
+                close_span.set_tag(websocket.CLOSE_REASON, reason)
```

Comment:
Same here. Can we create a trace handler to set this logic. This integration should Ideally just register and dispatch events. This will allow ASM and other products to add their own websocket support without modifying this integration file.